packages/dcos-integration-test/extra/test_legacy_user_management.py

```diff
@@ -0,0 +1,184 @@
+
+    # This is the old behavior in dcos-oauth.
+    # assert r.status_code == 500
+    # assert 'invalid email' in r.text
+
+    # With Bouncer non-email uids are valid, and the request fails as of the
+    # missing request body.
+    assert r.status_code == 400
+    assert 'Request has bad Content-Type or lacks JSON data' in r.text
+
+
+@pytest.mark.usefixtures('remove_users_added_by_test')
+def test_legacy_user_creation_with_empty_json_doc(dcos_api_session):
+    # Legacy HTTP clients built for dcos-oauth such as the web UI (up to DC/OS
+    # 1.12) might insert users in the following way: uid appears to be an email
+    # address, and the JSON document in the request body does not provide a
+    # `public_key` or a `password` property (indicating local host), or is
+    # empty. The web UI expects those users to be remote users, usable with the
+    # legacy OIDC ID Token login method through the 'https://dcos.auth0.com/'
+    # provider. This behavior is maintained in Bouncer for backwards
+    # compatibility.
+    r = dcos_api_session.put('/acs/api/v1/users/user@domain.foo', json={})
+    assert r.status_code == 201, r.text
+
+    # Bouncer annotates the created user (this is new compared to dcos-oauth).
+    r = dcos_api_session.get('/acs/api/v1/users/user@domain.foo')
+    assert r.json()['provider_type'] == 'oidc'
+    assert r.json()['provider_id'] == 'https://dcos.auth0.com/'
+    assert r.json()['is_remote'] is True
+
+    # When the uid however does not appear to be an email address the more sane
+    # behavior of Bouncer takes effect: an empty (meaningless) JSON body
+    # results in a useful error message.
+    r = dcos_api_session.put('/acs/api/v1/users/user1', json={})
+    assert r.status_code == 400
+    assert 'One of `password` or `public_key` must be provided' in r.text
+
+
+@pytest.mark.usefixtures('remove_users_added_by_test')
+def test_user_put_email_uid_and_description(dcos_api_session):
+    r = dcos_api_session.put(
+        '/acs/api/v1/users/user1@domain.foo',
+        json={'description': 'integration test user'}
+    )
+    assert r.status_code == 201, r.text
+
+    users = get_users(dcos_api_session)
+    assert len(users) > 1
+    assert 'user1@domain.foo' in users
+
+
+@pytest.mark.usefixtures('remove_users_added_by_test')
+def test_user_put_with_legacy_body(dcos_api_session):
+    # The UI up to DC/OS 1.12 sends the `creator_uid` and the `cluster_url`
+    # properties although they are not used by dcos-oauth. Bouncer supports
+    # these two properties for legacy reasons. Note(JP): As a follow-up task we
+    # should change the UI to not send these properties anymore, and then remove
+    # the properties from Bouncer's UserCreate JSON schema again, ideally within
+    # the 1.13 development cycle.
+    r = dcos_api_session.put(
+        '/acs/api/v1/users/user2@domain.foo',
+        json={'creator_uid': 'any@thing.bla', 'cluster_url': 'foobar'}
+    )
+    assert r.status_code == 201, r.text
+
```

Comment:
Should this have the check that `user2@domain.foo` has been added, as in the previous test?

Reply:
Well, these tests were most meaningful in the exact moment when I replaced dcos-oauth with bouncer.

I wrote this test module very quickly to understand a bit about dcos-oauth's behavior. Then I kept this test module to show how behavior changed subtly, or how certain behavior did not change.

Going forward, this test module is not meant to be extended.

The user management stuff in Bouncer is actually very properly tested in Bouncer's tests as well as in the DC/OS Enterprise integration test suite.

We can add more extensive and systematic user management tests to the new test_iam module.